### PR TITLE
update: support registering device-specific update handlers

### PR DESCRIPTION
**Pull request by anupcshan**
Useful to update devices which expect blobs at fixed locations

Related to #101

### CODE REVIEW

**Review by stapelberg — CHANGES_REQUESTED** [review]
Thanks for your PR!

I think it’d be better to only offer a function that sets up the extra update handlers based on the device type, instead of offering a function that allows to set arbitrary update handlers.

E.g., how about `gokrazy.SetupDeviceSpecifics()`, taking a constant identifying the system (e.g. odroid)?

**Comment by anupcshan** [comment]
> Thanks for your PR!
> 
> I think it’d be better to only offer a function that sets up the extra update handlers based on the device type, instead of offering a function that allows to set arbitrary update handlers.
> 
> E.g., how about `gokrazy.SetupDeviceSpecifics()`, taking a constant identifying the system (e.g. odroid)?

Hmmm, I'm trying to figure out where gokrazy process would get the list of device specific handlers, in that case.
As it is implemented now, `gokr-packer` is responsible for parsing the device manifest and passing along handler information to `gokrazy` via `main()` in the generated `init,go`.

There are a few other ways to achieve this:
1. Store relevant config in some kind of `map[deviceType]deviceDetails` in `gokrazy/gokrazy`. This way, we can get rid of the device manifest file altogether and pass along `mbr-but-not-gpt` and "device type" as flags to `gokr-packer`. Makes it easy to call `gokrazy.SetupDeviceSpecifics(deviceType)` via `init.go` in this approach. This is the simplest way to implement it. Sounds like this is what you're suggesting.
2. Keep the device manifest around, dump the manifest file in the root image and have it parsed and acted on by `gokrazy`. Instead of passing along arbitrary update handlers, `gokrazy` just reads the (optional) manifest on start and registers appropriate handlers.

What do you think?

**Comment by stapelberg** [comment]
> Hmmm, I'm trying to figure out where gokrazy process would get the list of device specific handlers, in that case.

What does https://pkg.go.dev/github.com/gokrazy/gokrazy#Model return on the odroid? (You can see the detected model in the gokrazy web interface.)

If the odroid devices report their model correctly, that might be the easiest way.

Keeping a map of device specifics inside gokrazy (option 1 in your reply) sounds preferable over option 2 to me.

**Comment by anupcshan** [comment]
> > Hmmm, I'm trying to figure out where gokrazy process would get the list of device specific handlers, in that case.
> 
> What does https://pkg.go.dev/github.com/gokrazy/gokrazy#Model return on the odroid? (You can see the detected model in the gokrazy web interface.)
> 
> If the odroid devices report their model correctly, that might be the easiest way.

They return "Hardkernel Odroid HC1" (from https://github.com/torvalds/linux/blob/c9e6606c7fe92b50a02ce51dda82586ebdf99b48/arch/arm/boot/dts/exynos5422-odroidhc1.dts#L14). That's unique enough to identify this device type.

> 
> Keeping a map of device specifics inside gokrazy (option 1 in your reply) sounds preferable over option 2 to me.

Yep, I've gone ahead and stored the device configs in `gokrazy/internal` - https://github.com/gokrazy/internal/pull/9. Its not in `gokrazy/gokrazy` because `gokr-packer` also needs to reference this device config while generating disk image/updating.

Let me make the change to use `gokrazy.Model` as the device type key (and rename that to device model).

**Comment by anupcshan** [comment]
OK, PRs for all relevant changes are ready now (in dependency order):
1. https://github.com/gokrazy/internal/pull/9 has the device config itself with some validation checks
2. This PR calls `gokrazy.SetupDeviceSpecifics` in `SuperviseServices` - uses `gokrazy.Model` to automatically register device-specific stuff.
3. https://github.com/gokrazy/tools/pull/28 to do the gokr-packer bits.

Each PR depends on the one before it (needs a go mod upgrade to pick up previous change).

I use https://github.com/anupcshan/odroidbake/blob/main/Makefile to manage a test installation.

File: update.go [hunk 1]
```diff
@@ -240,6 +286,11 @@ func initUpdate() error {
 	http.HandleFunc("/update/root", nonConcurrentUpdateHandler(rootdev.Partition(rootdev.InactiveRootPartition())))
 	http.HandleFunc("/update/switch", nonConcurrentSwitchHandler(rootdev.InactiveRootPartition()))
 	http.HandleFunc("/update/testboot", nonConcurrentTestbootHandler(rootdev.InactiveRootPartition()))
+
+	for _, extraUpdateHandler := range extraUpdateHandlers {
+		http.HandleFunc(fmt.Sprintf("/update/%s", extraUpdateHandler.name), nonConcurrentLimitedUpdateHandler(extraUpdateHandler.device, extraUpdateHandler.offset, extraUpdateHandler.maxLength))
```
Comment (stapelberg) [inline update.go:291]: I think an extra level of hierarchy would be good in the paths. How about `/update/device-specific/<name>`?

File: update.go [hunk 2]
```diff
@@ -225,6 +243,34 @@ func nonConcurrentTestbootHandler(newRootPartition int) func(http.ResponseWriter
...
+var extraUpdateHandlers []*ExtraUpdateHandler
+
+func setupDeviceSpecifics() {
+	modelName := Model()
+	cfg, ok := deviceconfig.DeviceConfigs[modelName]
+	if !ok {
+		return
+	}
+
+	log.Printf("setting up device-specific handlers for '%s'", modelName)
```
Comment (stapelberg) [inline update.go:262]: Use %q instead of '%s' here

Comment (stapelberg) [inline update.go:262]: Make it “device-specific update handlers” instead of “device-specific handlers”

File: update.go [hunk 3]
```diff
@@ -225,6 +243,34 @@ func nonConcurrentTestbootHandler(newRootPartition int) func(http.ResponseWriter
 	}
 }
 
+type ExtraUpdateHandler struct {
```
Comment (stapelberg) [inline update.go:246]: Why is this type exported? It doesn’t need to be, right? Can you change it to `extraUpdateHandler` (lower-case e)?

**Review by stapelberg — COMMENTED** [review]
This is looking much better now!

**Review by stapelberg — APPROVED** [review]

**Comment by stapelberg** [comment]
This PR is good to go! Can you remove the `replace` directive in go.mod and pull in the latest `gokrazy/internal` please?

**Comment by anupcshan** [comment]
> This PR is good to go! Can you remove the `replace` directive in go.mod and pull in the latest `gokrazy/internal` please?

Thanks for the thorough review! Updated.

**Review by stapelberg — APPROVED** [review]

**Pull request merged**